### PR TITLE
Add a function to retrieve the shape index from a leaf BVH node

### PR DESCRIPTION
**Pull request by h3r2tic**
Hey!

I'm using your crate to feed a BVH to the GPU with some custom packing/flattening. To do that, I needed to add a function to retrieve the shape index from a leaf node.

Cheers, and thanks for sharing your code!

### CODE REVIEW

**Comment by svenstaro** [comment]
Can you make a test for that?

**Comment by h3r2tic** [comment]
Yep, no problem.

File: src/bvh/bvh.rs [hunk 1]
```diff
@@ -200,6 +200,14 @@ impl BVHNode {
         }
     }
 
+    /// Returns the index of the shape contained in the node, assuming it is a leaf.
+    pub fn shape_index(&self) -> usize {
+        match *self {
+            BVHNode::Leaf { shape_index, .. } => shape_index,
+            _ => panic!("Tried to get the shape index of an interior node."),
```
Comment (svenstaro) [inline src/bvh/bvh.rs]: I'm not sure this should panic. Can't we rather define this to be `None` and return and `Option`?

Comment (h3r2tic) [reply src/bvh/bvh.rs]: Yeah, that does sound better. One sec!

Comment (svenstaro) [reply src/bvh/bvh.rs]: `panic!` should happen if you get run an invalid operation and I can't quite decide whether this is an invalid operation or whether it's actually valid to run this on non-`Leaf` nodes. What do you think?

**Comment by svenstaro** [comment]
By the way, how are you using this exactly? :)

File: src/bvh/bvh.rs (hunk 1 above)
Comment (h3r2tic) [reply src/bvh/bvh.rs]: I think Option is a bit prettier. I wasn't too fond of having to deconstruct the enum just to test whether it's a node or a leaf, *and then* having the `shape_index` fn do another runtime check of self's variant.

**Comment by svenstaro** [comment]
Also please make a test case for the `None` case. Should be easy. :)

**Comment by h3r2tic** [comment]
Done!

As for how I'm using it, I'm packing the BVH into a format that only uses one child index (the other one follows directly in memory), and doing stackless traversal on the GPU. And then I'm generating five more of those trees -- reusing nodes, only flattening differently. They're specialized for the ray traveling more or less in +X, -X, +Y, -Y, +Z, -Z directions. The intersection shader picks the correct tree at the start.

It's based on the [ideas of Toshiya Hachisuka](https://www.ci.i.u-tokyo.ac.jp/~hachisuka/tdf2015.pdf).

Here's some very rough work in progress if you're curious: https://github.com/h3r2tic/rtoy-samples/blob/master/src/bin/raytrace/main.rs
Along with the corresponding shader: https://github.com/h3r2tic/rtoy-samples/blob/master/assets/shaders/raytrace.glsl

**Comment by svenstaro** [comment]
Super sweet reference! I'm going to be using that for my next path tracer. I've already done a few in the past but didn't see that document yet.

File: src/bvh/bvh.rs [hunk 2]
```diff
@@ -200,6 +200,15 @@ impl BVHNode {
         }
     }
 
+    /// Returns the index of the shape contained within the node if is a leaf,
+    /// or None if it is an interior node.
```
Comment (svenstaro) [inline src/bvh/bvh.rs]: Tiny nitpick: `None` instead of None.

Comment (h3r2tic) [reply src/bvh/bvh.rs]: :+1:  fixed.

**Comment by h3r2tic** [comment]
Toshiya is a legend :) I'm a bit of an internet stalker of his.

**Comment by svenstaro** [comment]
I'll pop out 0.3.1 with your changes once travis is happy. Feel free to submit more stuff to this. :)

**Comment by h3r2tic** [comment]
Awesome, thanks! :)
